### PR TITLE
perf: optimize DeepSeek-V4

### PR DESCRIPTION
**Pull request by 0xClandestine**
## DeepSeek V4 Optimization Changelog

Targets the DeepSeek-V4/V4-Flash Metal kernels in `mlx_lm/models/deepseek_v4.py`.

### What stuck

| Change | Impact |
|---|---|
| **Vectorized HC Sinkhorn kernel** — Rewrote `_make_hc_split_sinkhorn_kernel` to use `float4` SIMD loads/stores, `dot()` for row sums, and fully unrolled 4x4 Sinkhorn iterations instead of scalar loops. | Measurable prefill speedup; approved by maintainer. |
| **`_overlap_transform` concatenate** — Replaced `mx.full` + scatter-assign with `mx.concatenate` of sliced tensors, avoiding a large intermediate allocation. | Approved by maintainer ("this is great"). |
| **Per-head q-norm Metal kernel** (`ds4_q_norm`) — Fused RMS norm for query vectors using a 32-thread SIMD group reduction, replacing `mx.rsqrt` + elementwise. | Additive, no regression. |
| **`hc_mult != 4` guard** — Falls back to pure-MLX Sinkhorn when `hc_mult` is not 4, since the float4 kernel is hardcoded for 4x4. | Safety fix. |

### What was tried and reverted

| Change | Why reverted |
|---|---|
| **Fused sparse attention Metal kernel** (`ds4_fused_sparse_attn`) — Online-softmax over local + sparse KV in a single MSL kernel. Multiple iterations fixing address spaces, dispatch, output shapes. | Hung prompt processing; maintainer benchmarked at 86 tok/s vs 122 tok/s prefill. |
| **`_split_sparse_attention`** (MLX fallback) — Two-pass local+sparse attention with log-sum-exp merge, `@mx.compile`-decorated. Replaced by standard SDPA with gather+concatenate. | Still in file as dead code but unused. SDPA path is faster. |
| **Fused partial RoPE Metal kernel** (`ds4_partial_rope`) — Eliminated split/concat intermediate for partial rotary embeddings. | Didn't move the needle on benchmarks. Reverted per maintainer feedback. |
| **Compressor x-buffer decode deferral** — Deferred `wkv`/`wgate` GEMVs until a full compression window was ready. | Broke batching. Reverted per maintainer feedback. |
| **HyperConnection `fn` in model dtype** — Changed `self.fn` from `float32` to default (bf16), split `flat`/`flat_f32` to do matmul in model dtype. | Caused prefill regression (78 tok/s down from 120). Reverted. |
| **SwitchGLU refactor** — Removed `scores` param, used `_gather_sort` helper, moved score multiplication outside to `DeepseekV4MoE.__call__`. | Caused prefill regression. Reverted to maintainer's approach. |
| **Simplified compress path** — Stripped `select_all` optimization, `lengths`/`pooled_mask` tracking, mask trimming, and `pooled_bias` handling. | Major prefill regression (78 → 262 tok/s after restoring). The `select_all` fast-path and mask trim are critical. |
| **Collapse/expand Metal kernels + MoE gate bf16** — Custom kernels for HyperConnection collapse/expand ops. | Reverted immediately (broke things). |
| **Tokenizer `PreTrainedTokenizerFast` fallback** — Caught `AttributeError` from `AutoTokenizer` for custom models with `rope_scaling`. | Upstream transformers fix (PR #45643) made it unnecessary. |

### Final benchmark

```
Prefill:    262.7 tok/s
Generation:  31.4 tok/s
Peak memory: 152.2 GB
```

### CODE REVIEW

File: mlx_lm/models/deepseek_v4.py [hunk 1]
```diff
@@ -244,15 +239,141 @@ def _rope_full(
     return pe_out
 
 
+def _make_partial_rope_kernel():
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py]: Doesn't affect perf, we can revert

File: mlx_lm/models/deepseek_v4.py [hunk 2]
```diff
@@ -300,97 +421,84 @@ def _make_hc_split_sinkhorn_kernel():
...
-
-        float pre_scale = static_cast<float>(scale[0]);
-        float post_scale = static_cast<float>(scale[1]);
-        float comb_scale = static_cast<float>(scale[2]);
-
-        for (int i = 0; i < HC; ++i) {
-            float z = static_cast<float>(mix[i]) * pre_scale
-                + static_cast<float>(base[i]);
-            pre_out[i] = 1.0f / (1.0f + metal::fast::exp(-z)) + epsv;
+        const auto mix = mixes + idx * MIX;
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py]: Great, but I believe we can imrpove perf here by removing looks but still using metal methods (i.e, metal::fast::exp(-z))) and SIMD

Comment (Blaizzy) [reply mlx_lm/models/deepseek_v4.py]: In other words, lets vectorize and ensure this utilizes maximum bandwidth

File: mlx_lm/models/deepseek_v4.py [hunk 3]
```diff
@@ -407,15 +515,17 @@ def hc_split_sinkhorn(
...
     if not isinstance(eps, mx.array):
         eps = mx.array([eps], dtype=mx.float32)
+    n_rows = mixes.size // ((2 + hc_mult) * hc_mult)
+    n_rows_arr = mx.array([n_rows], dtype=mx.int32)
     return _hc_split_sinkhorn_kernel(
-        inputs=[mixes, scale, base, eps],
+        inputs=[mixes, scale, base, eps, n_rows_arr],
         template=[("HC", hc_mult), ("ITERS", sinkhorn_iters)],
-        grid=(mixes.size // ((2 + hc_mult) * hc_mult), 1, 1),
+        grid=((n_rows + 255) & ~255, 1, 1),
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py]: Don't really affect perf in any way so we can revert

File: mlx_lm/models/deepseek_v4.py [hunk 4]
```diff
@@ -448,6 +558,222 @@ def _rms_rsqrt(flat: mx.array, eps: float) -> mx.array:
     return mx.rsqrt((flat * flat).mean(axis=-1, keepdims=True) + eps)
 
 
+def _make_fused_sparse_attn_kernel():
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py]: We can revert

File: mlx_lm/models/deepseek_v4.py [hunk 5]
```diff
@@ -457,15 +783,17 @@ def __init__(self, config: ModelArgs):
...
     def compute_weights(self, x: mx.array):
         B, L, H, D = x.shape
-        flat = x.reshape(B, L, H * D).astype(mx.float32)
-        rsqrt = _rms_rsqrt(flat, self.norm_eps)
-        mixes = (flat @ self.fn.T) * rsqrt
+        flat = x.reshape(B, L, H * D)
+        flat_f32 = flat.astype(mx.float32)
+        rsqrt = _rms_rsqrt(flat_f32, self.norm_eps)
+        # fn is bfloat16; match flat's dtype so MLX uses a native bf16 GEMV.
+        mixes = (flat.astype(self.fn.dtype) @ self.fn.T).astype(mx.float32) * rsqrt
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py]: I tried but can't tell the difference.

File: mlx_lm/models/deepseek_v4.py [hunk 6]
```diff
@@ -611,9 +941,8 @@ class DeepseekV4Cache:
 
     def __init__(self, sliding_window: int):
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py:844]: The changes here affect batching

File: mlx_lm/models/deepseek_v4.py [hunk 7]
```diff
@@ -1213,10 +1428,12 @@ def __init__(self, config: ModelArgs, compress_ratio: int, head_dim: int):
...
-        out = mx.full((B, W, 2 * R, self.head_dim), fill_value, dtype=x.dtype)
-        out[:, :, R:] = x[:, :, :, self.head_dim :]
-        out[:, 1:, :R] = x[:, :-1, :, : self.head_dim]
-        return out
+        second_half = x[:, :, :, self.head_dim :]  # (B, W, R, head_dim)
+        fill_row = mx.full((B, 1, R, self.head_dim), fill_value, dtype=x.dtype)
+        prev_first = mx.concatenate(
+            [fill_row, x[:, :-1, :, : self.head_dim]], axis=1
+        )  # (B, W, R, head_dim)
+        return mx.concatenate([prev_first, second_half], axis=2)  # (B, W, 2R, head_dim)
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py:1454]: This is great, but make sure concurrent requests don't get affect on the server.

File: mlx_lm/models/deepseek_v4.py [hunk 8]
```diff
@@ -1269,6 +1495,11 @@ def __call__(
 
 
 class Indexer(nn.Module):
+    # NOTE: Hadamard rotation omitted (see Compressor note above).
+    # NOTE: FP4 query simulation omitted. The HF reference quantizes Q to FP4
+    # before scoring compressed KV for training/inference parity. Running with
+    # full-precision queries gives slightly different top-k selections on edge
+    # cases, but the top-k selection is robust enough that this rarely matters.
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py]: I would remove

File: mlx_lm/models/deepseek_v4.py [hunk 9]
```diff
@@ -1315,6 +1546,75 @@ def __call__(
         return mx.argpartition(-scores, kth=k - 1, axis=-1)[..., :k]
 
 
+@mx.compile
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py]: This is the main culprit for slowing down prefill, i suspect that it can be optimized with but the question is what's the impact on the overall performance if it's working well?

File: mlx_lm/models/deepseek_v4.py [hunk 10]
```diff
@@ -1645,9 +1898,16 @@ def predicate(k):
...
+                    (
+                        ".attn_hc.base",
+                        ".attn_hc.scale",
+                        ".ffn_hc.base",
+                        ".ffn_hc.scale",
+                        ".hc_head.base",
+                        ".hc_head.scale",
+                    )
+                )
             )
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py]: We can revert

File: mlx_lm/tokenizer_utils.py [hunk 11]
```diff
@@ -611,9 +611,18 @@ def load(
...
+        tokenizer = AutoTokenizer.from_pretrained(
+            model_path, **(tokenizer_config_extra or {})
+        )
+    except AttributeError:
+        # Custom model types (e.g. deepseek_v4) may have config fields like
+        # rope_scaling that trigger transformers' config standardization which
+        # expects attributes (max_position_embeddings) absent from the generic
+        # PreTrainedConfig.  Fall back to loading the tokenizer directly.
+        tokenizer = PreTrainedTokenizerFast.from_pretrained(
+            model_path, **(tokenizer_config_extra or {})
```
Comment (Blaizzy) [inline mlx_lm/tokenizer_utils.py]: We can revert, because the TFMS PR solves this: https://github.com/huggingface/transformers/pull/45643

**Review by Blaizzy — COMMENTED** [review]
Thanks a lot! Left a few comments

File: mlx_lm/tokenizer_utils.py (hunk 11 above)
Comment (0xClandestine) [reply mlx_lm/tokenizer_utils.py]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 10 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 9 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 8 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 7 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py:1454]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 6 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py:844]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 5 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 4 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 3 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 2 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py (hunk 1 above)
Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py]: [9126d7c](https://github.com/Blaizzy/mlx-lm/pull/13/commits/9126d7c8f51f51a6d8a247a4a3468b385fc5a8ea)

File: mlx_lm/models/deepseek_v4.py [hunk 12]
```diff
@@ -1645,9 +1696,6 @@ def predicate(k):
             return not (
                 "attn_sink" in k
```
Comment (Blaizzy) [inline mlx_lm/models/deepseek_v4.py:1646]: revert this deletion

Comment (0xClandestine) [reply mlx_lm/models/deepseek_v4.py:1646]: [a3c04af](https://github.com/Blaizzy/mlx-lm/pull/13/commits/a3c04af7f5653680859f5ed923c82f0a35a624ad)

File: README.md
Comment (Blaizzy) [inline README.md:1]: Let’s remove the readme changes for now

**Review by Blaizzy — APPROVED** [review]
LGTM! 🚀
